### ISSUE TITLE
Can we change `bytes` to use `char`/`unsigned char`/`std::byte` instead of `int8_t`?

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#14689]
Why is `bytes` defined like this?
```cpp
using bytes = basic_sstring<int8_t, uint32_t, 31, false>;
```

It's a PITA to work with this -- for example we cannot construct it from an array of `char`s because that would require `reinterpret_cast<int8_t*>` which breaks type aliasing rules.

Can we change it to use `std::byte`?

**Comment by kbr-scylla** [comment]
This PR and my review of it was what triggered the question:
https://github.com/scylladb/scylladb/pull/14534#discussion_r1262569389

**Comment by kbr-scylla** [comment]
> Can we change it to use std::byte?

https://en.cppreference.com/w/cpp/types/byte

**Comment by michoecho** [comment]
> Why is bytes defined like this?

30b40bf7b1ec6bce3240032724b098a0ae554997

And technically it's more optimizer-friendly *because* it's a distinct type from `char`. When you write something to a `int8_t` buffer (through a pointer), the optimizer can assume that it doesn't overlap any other pointers due to type aliasing rules. Otherwise it would have to re-load the contents of all pointers which were loaded before the int8_t write. But: I'm not sure if mainstream compilers exploit that (I'd guess they don't), and I'm not sure if it makes any performance difference (I'd guess it doesn't).

So I'd say: go for it. But maybe Avi doesn't like the idea.

> It's a PITA to work with this -- for example we cannot construct it from an array of chars because that would require reinterpret_cast<int8_t*> which breaks type aliasing rules.

Oh, my sweet summer child:
https://en.cppreference.com/w/cpp/named_req/ImplicitLifetimeType
https://en.cppreference.com/w/cpp/language/object#Object_creation

It doesn't break type aliasing, because char arrays aren't arrays of char. They are a quantum superposition of all combinations of all implicit-lifetime type objects, until you observe them through a non-omnipotent pointer.

And what if you observe them through two differently-typed non-omnipotent pointers to implicit-lifetime types? I don't dare to answer what happens.

**Comment by kbr-scylla** [comment]
> It doesn't break type aliasing, because char arrays aren't arrays of char. They are a quantum superposition of all combinations of all implicit-lifetime type objects, until you observe them through a non-omnipotent pointer.

I have no idea what you just said.

**Comment by kbr-scylla** [comment]
From https://en.cppreference.com/w/cpp/language/reinterpret_cast
> Any object pointer type T1* can be converted to another object pointer type cv T2*. (...) In any case, the resulting pointer may only be dereferenced safely if allowed by the type aliasing rules (see below). 

And the type aliasing rules:
> Whenever an attempt is made to read or modify the stored value of an object of type DynamicType through a glvalue of type AliasedType, the behavior is undefined unless one of the following is true:
> 
> - AliasedType and DynamicType are similar.
> - AliasedType is the (possibly cv-qualified) signed or unsigned variant of DynamicType.
> - AliasedType is std::byte, (since C++17) char, or unsigned char: this permits examination of the object representation of any object as an array of bytes. 
> 
> Informally, two types are similar if, ignoring top-level cv-qualification:
> 
> - they are the same type; or
> - they are both pointers, and the pointed-to types are similar; or
> - they are both pointers to member of the same class, and the types of the pointed-to members are similar; or 
> - they are both arrays of the same size or both arrays of unknown bound, and the array element types are similar. | (until C++20)
> - they are both arrays of the same size or at least one of them  is array of unknown bound, and the array element types are similar. | (since C++20)



So if I understand correctly, `DynamicType` in our case is `char` and `AliasedType` is `int8_t`. And this case is not covered by the type aliasing rules.

**Comment by michoecho** [comment]
> So if I understand correctly, DynamicType in our case is char

No, the dynamic type is int8_t (which is an implicit-lifetime type), because that what's needed to prevent undefined behaviour. An array of `int8_t` has been _implicitly created_ inside the `char` storage, you just don't know it.

> Objects of [implicit-lifetime types](https://en.cppreference.com/w/cpp/named_req/ImplicitLifetimeType) can also be implicitly created by
> ... 
> - operations that begin lifetime of an array of type unsigned char or [std::byte](https://en.cppreference.com/w/cpp/types/byte) (since C++17), in which case such objects are created in the array,
> ...
>
> Zero or more objects may be created in the same region of storage, as long as doing so would give the program defined behavior. If such creation is impossible, e.g. due to conflicting operations, the behavior of the program is undefined. If multiple such sets of implicitly created objects would give the program defined behavior, it is unspecified which such set of objects is created. In other words, implicitly created objects are not required to be uniquely defined.

**Comment by michoecho** [comment]
> which line contains this operation? Is it the arr constructor (i.e. first line)? This is where an array of int8_t is implicitly created?

Yes.

> Another question -- what if we're dealing with std::array<char, 32>? It looks like the rule doesn't apply there after all.

True, it doesn't apply. It used to (or at least I thought it did), but 5 months ago it was declared a defect in the standard.

> Does the rule apply in this case as well? Can we safely consider &str[0] to be "an array of chars"?

Yes.

> Even if we can -- those are chars, not unsigned chars, so the rule still doesn't seem to apply?

True, the rule doesn't apply.

But the standard rules are a lie, anyway. In C, `signed char` and `char` do have the magic aliasing properties, so no compiler is insane enough to exploit the C++ rule during optimization.

See:
> https://github.com/gcc-mirror/gcc/blob/da93c41c4ea124d61d41fb8629b801f74886284c/gcc/c-family/c-common.cc#L3820
> https://github.com/llvm/llvm-project/blob/3cd3f11c174baa001b337b88c7a6507eb5705cf2/clang/lib/CodeGen/CodeGenTBAA.cpp#L121-L130

So your example is technically illegal, but good luck finding a compiler where the undefined behavior actually causes something unexpected.

**Comment by kostja** [comment]
> > which line contains this operation? Is it the arr constructor (i.e. first line)? This is where an array of int8_t is implicitly created?
> 
> Yes.
> 
> > Another question -- what if we're dealing with std::array<char, 32>? It looks like the rule doesn't apply there after all.
> 
> True, it doesn't apply. It used to (or at least I thought it did), but 5 months ago it was declared a defect in the standard.
> 
> > Does the rule apply in this case as well? Can we safely consider &str[0] to be "an array of chars"?
> 
> Yes.
> 
> > Even if we can -- those are chars, not unsigned chars, so the rule still doesn't seem to apply?
> 
> True, the rule doesn't apply.
> 
> But the standard rules are a lie, anyway. In C, `signed char` and `char` do have the magic aliasing properties, so no compiler is insane enough to exploit the C++ rule during optimization.
> 
> See:
> 
> > https://github.com/gcc-mirror/gcc/blob/da93c41c4ea124d61d41fb8629b801f74886284c/gcc/c-family/c-common.cc#L3820
> > https://github.com/llvm/llvm-project/blob/3cd3f11c174baa001b337b88c7a6507eb5705cf2/clang/lib/CodeGen/CodeGenTBAA.cpp#L121-L130
> 
> So your example is technically illegal, but good luck finding a compiler where the undefined behavior actually causes something unexpected.

Exactly, C provides some level of sanity and stability rendering C++ rules irrelevant. In the end chipmakers will just move on and  this whole story will become legacy.

**Comment by avikivity** [comment]
> Why is `bytes` defined like this?
> 
> ```c++
> using bytes = basic_sstring<int8_t, uint32_t, 31, false>;
> ```


Compatibility with Cassandra, and intended incompatibility with string_view or other APIs that talk about strings.

**Comment by nyh** [comment]
> > Why is `bytes` defined like this?
> > ```c++
> > using bytes = basic_sstring<int8_t, uint32_t, 31, false>;
> > ```
> 
> Compatibility with Cassandra

Right, if you try changing it to uint8_t  hopefully (?) we would have CQL tests that start to fail when clustering keys are sorted in the wrong order, or something.

> and intended incompatibility with string_view or other APIs that talk about strings.

This is a good point. We often use "bytes" for things like encoded keys (multiple components) and don't want to confuse it with a single string key. So it's good to have efficient conversions between bytes and strings, but perhaps less good to accidentally have implicit conversions between the two.
